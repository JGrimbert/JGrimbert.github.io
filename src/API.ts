const API_URL = 'http://henri-potier.xebia.fr/books/';

export const API = ({

    getBooks: async () => (await fetch(API_URL)).json(),

    getOffers: async (url: string[]) => {
        return url.length && (await fetch(`${API_URL}${url}/commercialOffers`)).json();
    }

});
