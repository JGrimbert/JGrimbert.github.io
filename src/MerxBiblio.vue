<script setup lang="ts">
  import { onMounted } from 'vue'
  import { biblio } from "./model";
  import { API } from "./API";
  import MerxBiblioDiscount from './components/MerxBiblioDiscount.vue'
  import MerxBiblioBag from './components/MerxBiblioBag.vue'
  import MerxBiblioProposal from './components/MerxBiblioProposal.vue'

  onMounted(async () => {
    biblio.complete = await API.getBooks();
  });
</script>

<template>
  <main>
      <MerxBiblioProposal />
      <MerxBiblioDiscount />
      <MerxBiblioBag />
  </main>
</template>

<style scoped>
main {
  display: flex;
  flex: 1;
  flex-direction: column;
}
</style>
