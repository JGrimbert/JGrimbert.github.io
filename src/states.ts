import {reactive, watchEffect} from "vue";
import {IBiblio, IDiscounts, IMerx} from "./model";
import {API} from "./API";

export const merx: IMerx = reactive({
    offers: [],
    initial: 0,
    discounted: 0
});

export const biblio: IBiblio = reactive({
    complete: [],
    selected: [],
    choosen: [],
    searched: "",
});

export const discounts: IDiscounts = {
    percentage: {
        html: ({ value }) => `Profitez d'une remise de ${value}%`,
        funk: ({ value, discount}) => value - (value/100)*discount,
    },
    minus: {
        html: ({ value }) => `Économisez directement ${value}€ en caisse`,
        funk: ({ value, discount}) => value - discount,
    },
    slice: {
        html: ({ value, sliceValue }) => `Vous pourriez bénéficier de la réduction de ${value}€ par tranche de ${sliceValue}€ d'achat`,
        funk: ({ value, discount, sliceValue}) => value - Math.floor(value/sliceValue) * discount,
    }
}

watchEffect(async () => {
    const { choosen, url, initial } = newOfferProposition(biblio.selected);
    /* (୨୧ ❛ᴗ❛) Set merx attributes */
    merx.offers = (await API.getOffers(url))?.offers;
    merx.initial = initial;
    merx.discounted = sumDiscounts();
    /* (୨୧ ❛ᴗ❛) Set choosen biblio */
    biblio.choosen = choosen;
});

function newOfferProposition (arr: any[], initial = 0) {

    const choosen = arr ? arr.reduce((acc, cur) => {
        initial = initial + cur.price
        return { ...acc, [cur.isbn]: cur }
    }, {}) : {}

    return { choosen, initial, url: Object.keys(choosen) };

}

function sumDiscounts() {
    return merx.offers && merx.offers.reduce((acc, cur) => {

        return discounts[cur.type].funk({
            value:acc,
            discount:cur.value,
            sliceValue:cur.sliceValue
        });

    }, merx.initial)
}
