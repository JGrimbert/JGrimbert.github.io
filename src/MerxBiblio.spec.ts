import { shallowMount } from '@vue/test-utils'
import { biblio, merx } from './states';
import MerxBiblio from './MerxBiblio.vue'
import { API } from "./API";

describe('MerxBiblio', () => {

    const mountBiblio = (biblio = [{}]) => {
        API.getBooks = jest.fn(() => Promise.resolve(biblio));
        return shallowMount(MerxBiblio);
    }

    it(`should fetch book list on mounted and update Biblio & Merx`, async () => {
        await mountBiblio([{}, {}]);
        expect(API.getBooks).toHaveBeenCalledTimes(1);
        expect(biblio).toHaveProperty("complete", [{},{}]);
        expect(biblio).toHaveProperty("choosen", {});
        expect(biblio).toHaveProperty("searched", "");
        expect(biblio).toHaveProperty("selected", []);
        expect(merx).toHaveProperty("offers", undefined);
        expect(merx).toHaveProperty("initial",0);
        expect(merx).toHaveProperty("discounted",undefined);
    });

    it(`should contain the proposal, discount and bag components`, async () => {
        const wrapper = await mountBiblio();
        expect(wrapper.html()).toMatchSnapshot();
    });
})
