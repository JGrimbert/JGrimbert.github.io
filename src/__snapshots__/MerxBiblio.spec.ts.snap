// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerxBiblio should contain the proposal, discount and bag components 1`] = `
"<main>
  <merx-biblio-proposal-stub></merx-biblio-proposal-stub>
  <merx-biblio-discount-stub></merx-biblio-discount-stub>
  <merx-biblio-bag-stub></merx-biblio-bag-stub>
</main>"
`;
